dynamiqs/time_array.py

```diff
@@ -288,158 +284,110 @@ def __init__(self, f: callable[[float], Array], f0: Array):
+        f = lambda t: self.f(t) * y
+        f0 = self.f0 * y
         return CallableTimeArray(f, f0)
 
-    def __add__(self, other: Array | TimeArray) -> TimeArray:
-        if isinstance(other, Array):
-            f = lambda t: self.f(t) + other
-            f0 = self.f0 + other
+    def __add__(self, y: ArrayLike | TimeArray) -> TimeArray:
+        if isinstance(y, get_args(ArrayLike)):
+            f = lambda t: self.f(t) + y
+            f0 = self.f0 + y
+            return CallableTimeArray(f, f0)
+        elif isinstance(y, ConstantTimeArray):
+            f = lambda t: self.f(t) + y.x
+            f0 = self.f0 + y.x
             return CallableTimeArray(f, f0)
-        elif isinstance(other, ConstantTimeArray):
-            return self + other.array
-        elif isinstance(other, CallableTimeArray):
-            f = lambda t: self.f(t) + other.f(t)
-            f0 = self.f0 + other.f0
+        elif isinstance(y, CallableTimeArray):
+            f = lambda t: self.f(t) + y.f(t)
+            f0 = self.f0 + y.f0
             return CallableTimeArray(f, f0)
         else:
             return NotImplemented
 
 
-class _PWCFactor:
+class _Factor:
+    @property
+    @abstractmethod
+    def shape(self) -> tuple[int, ...]:
+        pass
+
+    @abstractmethod
+    def conj(self) -> _Factor:
+        pass
+
+    @abstractmethod
+    def __call__(self, t: Scalar) -> Array:
+        pass
+
+    @abstractmethod
+    def reshape(self, *args: int) -> _Factor:
+        pass
+
+
+class _PWCFactor(_Factor):
     # Defined by a tuple of 2 arrays (times, values), where
-    # - times: (nv+1) are the time points between which the PWC array take constant
+    # - times: (nv+1) are the time points between which the PWC factor takes constant
     #          values, where nv is the number of time intervals
-    # - values: (..., nv) are the constant values for each time interval, where
-    #           (...) is an arbitrary batching size
+    # - values: (..., nv) are the constant values for each time interval, where (...)
+    #           is an arbitrary batching size
 
     def __init__(self, times: Array, values: Array):
         self.times = times  # (nv+1)
         self.values = values  # (..., nv)
         self.nv = self.values.shape[-1]
 
     @property
-    def shape(self) -> Tuple[int, ...]:
+    def shape(self) -> tuple[int, ...]:
         return self.values.shape[:-1]  # (...)
 
-    def conj(self) -> _PWCFactor:
+    def conj(self) -> _Factor:
         return _PWCFactor(self.times, self.values.conj())
 
-    def __call__(self, t: float) -> Array:
+    def __call__(self, t: Scalar) -> Array:
         if t < self.times[0] or t >= self.times[-1]:
             return jnp.zeros_like(self.values[..., 0])  # (...)
         else:
             # find the index $k$ such that $t \in [t_k, t_{k+1})$
             idx = jnp.searchsorted(self.times, t, side='right') - 1
             return self.values[..., idx]  # (...)
```

Comment:
I think this needs to be modified with some `lax` function to account for the if statement with the jit.

Reply:
Correct, let's fix JIT-stuff in another PR if you're ok with that.